[WIP] move: check PoW into AcceptBlock (from AcceptBlockHeader)

decryp2kanon opened the pull request:
# Move to: https://github.com/sugarchain-project/sugarchain/pull/122

# TL;DR
- CURRENT: 
  * check header during download, and sync later.
  * bottle neck at first stage
- PR:
  * download header first, and check later together when sync.
  * bottle neck gone! :tada: 
- AIM:
  * remove the bottle neck with moving the checking point.
  * do not accept fake header when attacked. (same expected react as before)

# RESULT
total IBD `50%` faster than moonlight
- :heavy_check_mark: header downloading `++700%` faster
- :x: sync block `--20%` slower (because check PoW moved to this stage)

![image](https://user-images.githubusercontent.com/60179867/80495034-56ab5e00-89a2-11ea-942b-6e0709679c96.png)

total download is around `4 GB`
- :heavy_check_mark: data traffic `50%` reduced
- actual blockchain is around `2.7 GB` 
![image](https://user-images.githubusercontent.com/60179867/80495635-20baa980-89a3-11ea-9502-ad2cdef7317b.png)


# Background
Unlike BTC (10 minutes block time), I thought in case 5 seconds blocktime, current PoW check is at **wrong position**. 

# What I did?
Therefore I moved the PoW check location from `AcceptBlockHeader` to `AcceptBlock` to remove a huge downloading bottle neck.

# Changes
BTC's IBD process is following:
- download HEADERS
  * check each headers PoW << BOTTLE NECK makes download slower...
- sync Blocks
  * download blocks

This PR is:
- download HEADERS << bottle neck is gone!
- sync Blocks
  * check each headers PoW
  * download blocks

FYI:
- HEADER = SHA256d
- PoW = YespowerSugar

# DEBUG
check header seems working. `CheckBlockH` called twice, when `AcceptBlock` ran.
- CheckBlockH: `2x`
- AcceptBlock: `1x`

- PRINTF
```
2020-04-27 21:41:46 CheckBlockH=63c0e66dd07b310f8f87ef95c625fa066188e3e9fc522d8139724897bb1819df
2020-04-27 21:41:46 CheckBlockH=63c0e66dd07b310f8f87ef95c625fa066188e3e9fc522d8139724897bb1819df
2020-04-27 21:41:46 AcceptBlock=63c0e66dd07b310f8f87ef95c625fa066188e3e9fc522d8139724897bb1819df
2020-04-27 21:41:46 CheckBlockH=4ce57fb82f4fd70b198100757ed8766938b02e07b2751b7c60bb17991a7654b5
2020-04-27 21:41:46 CheckBlockH=4ce57fb82f4fd70b198100757ed8766938b02e07b2751b7c60bb17991a7654b5
2020-04-27 21:41:46 AcceptBlock=4ce57fb82f4fd70b198100757ed8766938b02e07b2751b7c60bb17991a7654b5
2020-04-27 21:41:46 CheckBlockH=7a9a720f22ce62da3adb59308a923c47e1f3ffadac7c1b77a848bbb327893b10
2020-04-27 21:41:46 CheckBlockH=7a9a720f22ce62da3adb59308a923c47e1f3ffadac7c1b77a848bbb327893b10
2020-04-27 21:41:46 AcceptBlock=7a9a720f22ce62da3adb59308a923c47e1f3ffadac7c1b77a848bbb327893b10
```

- DEBUG.LOG
```
2020-04-27 21:41:46 UpdateTip: new best=63c0e66dd07b310f8f87ef95c625fa066188e3e9fc522d8139724897bb1819df height=419576 version=0x20000000 log2_work=40.518294 tx=486123 date='2019-09-17 08:07:03' progress=0.106275 cache=0.0MiB(3txo)
2020-04-27 21:41:46 UpdateTip: new best=4ce57fb82f4fd70b198100757ed8766938b02e07b2751b7c60bb17991a7654b5 height=419577 version=0x20000000 log2_work=40.518295 tx=486124 date='2019-09-17 08:07:04' progress=0.106275 cache=0.0MiB(6txo)
2020-04-27 21:41:46 UpdateTip: new best=7a9a720f22ce62da3adb59308a923c47e1f3ffadac7c1b77a848bbb327893b10 height=419578 version=0x20000000 log2_work=40.518297 tx=486126 date='2019-09-17 08:07:23' progress=0.106276 cache=0.0MiB(49txo)
```

# TEST CODE
- FAKE HEADER https://github.com/sugarchain-project/sugarchain/pull/119/commits/990d78bae7054175cd2419dc314e9235ef55f351
```cpp
    // FAKE HEADER POW TEST
    // uint256 fake_target_1 = uint256S("0x7d5eaec2dbb75f99feadfa524c78b7cabc1d8c8204f79d4f3a83381b811b0adc"); // block 0
    uint256 fake_target_2 = uint256S("0xce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91"); // block 1
    uint256 fake_target_3 = uint256S("0x72e36f3fcdf98d3625dfe03f28a914c513b913231e479d53fc22e5e46cf5b585"); // block 17
    uint256 fake_target_4 = uint256S("0xf6f565e58812f89d8ea2aca296b9934ea82918f5bd443312af90b0be1465dbd1"); // block 510
    uint256 fake_replacement = uint256S("0x7777777777777777777777777777777777777777777777777777777777777777");
    if (block.GetHash() == fake_target_2 || block.GetHash() == fake_target_3 || block.GetHash() == fake_target_4) {
        // int nHeight = pindexPrev->nHeight + 1;
        printf("%s INPUT FAKE HEADER! %s\n", DateTimeStrFormat("%Y-%m-%d %H:%M:%S", GetTime()).c_str(), block.GetHash().ToString().c_str());
        if (fCheckPOW && !CheckProofOfWork(fake_replacement, block.nBits, consensusParams)) {
            return state.DoS(50, false, REJECT_INVALID, "high-hash (fake)", false, "proof of work failed (fake)");
        }
    } else if (fCheckPOW && !CheckProofOfWork(block.GetPoWHash_cached(), block.nBits, consensusParams)) {
        return state.DoS(50, false, REJECT_INVALID, "high-hash", false, "proof of work failed");
    }
```

# TEST RESULT
- if you get some fake headers, daemon will bann the fake node. OK :heavy_check_mark: 
- if you recompile or delete `banlist.dat`, it syncs continuously well. OK :heavy_check_mark: 
![image](https://user-images.githubusercontent.com/60179867/80530309-ca199380-89d3-11ea-81ac-598c6786c75d.png)


- `printf`
```
2020-04-28 20:21:27 CheckBlockH=c62f3395a8c53ebedb637bf80c83239e7bcd465133ef8c911a16c116ea9f3602
2020-04-28 20:21:27 AcceptBlock=c62f3395a8c53ebedb637bf80c83239e7bcd465133ef8c911a16c116ea9f3602
2020-04-28 20:21:27 INPUT FAKE HEADER! ce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91
2020-04-28 20:21:27 INPUT FAKE HEADER! 72e36f3fcdf98d3625dfe03f28a914c513b913231e479d53fc22e5e46cf5b585
2020-04-28 20:21:28 INPUT FAKE HEADER! ce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91
2020-04-28 20:21:28 INPUT FAKE HEADER! 72e36f3fcdf98d3625dfe03f28a914c513b913231e479d53fc22e5e46cf5b585
```

- `debug.log`
```
2020-04-28 20:18:47 New outbound peer connected: version: 70015, blocks=4298257, peer=31
2020-04-28 20:21:26 Misbehaving: 216.155.157.233:34230 peer=1 (0 -> 50)
2020-04-28 20:21:26 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:26 Misbehaving: 216.155.157.233:34230 peer=1 (50 -> 100) BAN THRESHOLD EXCEEDED
2020-04-28 20:21:26 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:27 Misbehaving: 211.144.120.164:34230 peer=23 (0 -> 50)
2020-04-28 20:21:27 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:27 Misbehaving: 211.144.120.164:34230 peer=23 (50 -> 100) BAN THRESHOLD EXCEEDED
2020-04-28 20:21:27 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:27 Misbehaving: 45.77.192.211:34230 peer=0 (0 -> 50)
2020-04-28 20:21:27 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:27 Misbehaving: 45.77.192.211:34230 peer=0 (50 -> 100) BAN THRESHOLD EXCEEDED
```

- banned:
![image](https://user-images.githubusercontent.com/60179867/80534654-8bd3a280-89da-11ea-8202-6c19f4421f55.png)


# TODO:
- :white_check_mark:  ~~make test_bitcoin (pow or...) https://github.com/sugarchain-project/sugarchain/pull/119/commits/990d78bae7054175cd2419dc314e9235ef55f351~~
- :white_check_mark: ~~try test with revert this PR cee8630 (testing old case)~~
- short bantime `/120`
- revert disabled many headers https://github.com/sugarchain-project/sugarchain/pull/78

decryp2kanon committed:
move: check PoW into AcceptBlock (from AcceptBlockHeader)

decryp2kanon committed:
printf

decryp2kanon commented:
![image](https://user-images.githubusercontent.com/60179867/80419525-451b7500-8914-11ea-9be7-0c4eda25bf28.png)

decryp2kanon committed:
uint256 hash into if

decryp2kanon committed:
fix: printf

decryp2kanon committed:
printf

decryp2kanon commented:
check header seems working. `CheckBlockH` called twice, when `AcceptBlock` ran.
- CheckBlockH: `2x`
- AcceptBlock: `1x`

```
2020-04-27 21:41:46 CheckBlockH=63c0e66dd07b310f8f87ef95c625fa066188e3e9fc522d8139724897bb1819df
2020-04-27 21:41:46 CheckBlockH=63c0e66dd07b310f8f87ef95c625fa066188e3e9fc522d8139724897bb1819df
2020-04-27 21:41:46 AcceptBlock=63c0e66dd07b310f8f87ef95c625fa066188e3e9fc522d8139724897bb1819df
2020-04-27 21:41:46 CheckBlockH=4ce57fb82f4fd70b198100757ed8766938b02e07b2751b7c60bb17991a7654b5
2020-04-27 21:41:46 CheckBlockH=4ce57fb82f4fd70b198100757ed8766938b02e07b2751b7c60bb17991a7654b5
2020-04-27 21:41:46 AcceptBlock=4ce57fb82f4fd70b198100757ed8766938b02e07b2751b7c60bb17991a7654b5
2020-04-27 21:41:46 CheckBlockH=7a9a720f22ce62da3adb59308a923c47e1f3ffadac7c1b77a848bbb327893b10
2020-04-27 21:41:46 CheckBlockH=7a9a720f22ce62da3adb59308a923c47e1f3ffadac7c1b77a848bbb327893b10
2020-04-27 21:41:46 AcceptBlock=7a9a720f22ce62da3adb59308a923c47e1f3ffadac7c1b77a848bbb327893b10
```

```
2020-04-27 21:41:46 UpdateTip: new best=63c0e66dd07b310f8f87ef95c625fa066188e3e9fc522d8139724897bb1819df height=419576 version=0x20000000 log2_work=40.518294 tx=486123 date='2019-09-17 08:07:03' progress=0.106275 cache=0.0MiB(3txo)
2020-04-27 21:41:46 UpdateTip: new best=4ce57fb82f4fd70b198100757ed8766938b02e07b2751b7c60bb17991a7654b5 height=419577 version=0x20000000 log2_work=40.518295 tx=486124 date='2019-09-17 08:07:04' progress=0.106275 cache=0.0MiB(6txo)
2020-04-27 21:41:46 UpdateTip: new best=7a9a720f22ce62da3adb59308a923c47e1f3ffadac7c1b77a848bbb327893b10 height=419578 version=0x20000000 log2_work=40.518297 tx=486126 date='2019-09-17 08:07:23' progress=0.106276 cache=0.0MiB(49txo)
```

decryp2kanon committed:
try check

decryp2kanon committed:
comment

decryp2kanon committed:
comment

decryp2kanon commented:
TODO:
- same test on BTC 0.16.3
- make test_bitcoin (pow or...)
- try revert cee8630

decryp2kanon commented:
# 38-move 10:30
![image](https://user-images.githubusercontent.com/60179867/80469898-2e106d80-897c-11ea-97e0-df070a36407d.png)
![image](https://user-images.githubusercontent.com/60179867/80470083-7465cc80-897c-11ea-83eb-8590e890894b.png)

# 39-move 50% faster than moonlight
![image](https://user-images.githubusercontent.com/60179867/80488088-c87eaa00-8998-11ea-88e0-0e81a0106c86.png)

decryp2kanon commented:
# OFFTOPIC
revert: additional download headers during IBD
https://github.com/sugarchain-project/sugarchain/commit/cee863077020e80c870326ef4ca5009e38697c1f
- original: `if (nCount == MAX_HEADERS_RESULTS) {`
  * so fast, but traffic too much...
![image](https://user-images.githubusercontent.com/60179867/80499008-66797100-89a7-11ea-9ee2-b6f52bb9a551.png)

decryp2kanon commented:
similar question. however we do not skip, but just re-locate the point.
https://bitcointalk.org/index.php?topic=935585.0

decryp2kanon commented:
i made a bad PoW test and it seems working.
![image](https://user-images.githubusercontent.com/60179867/80519873-9e8ead00-89c3-11ea-94ce-b3a7e8f61e6e.png)

```cpp
    // Check proof of work matches claimed amount
    if (fCheckPOW && !CheckProofOfWork(uint256S("07eaa2805e4d72e523003ae98f01bb0059f6e4031747cb1b9060c9eb4443af7b"), block.nBits, consensusParams))
        return state.DoS(50, false, REJECT_INVALID, "high-hash", false, "proof of work failed");
```

OK

decryp2kanon committed:
try: fake PoW test

decryp2kanon commented:
![image](https://user-images.githubusercontent.com/60179867/80530309-ca199380-89d3-11ea-81ac-598c6786c75d.png)

# TEST: FAKE HEADER
```cpp
    // FAKE HEADER POW TEST
    // uint256 fake_target_1 = uint256S("0x7d5eaec2dbb75f99feadfa524c78b7cabc1d8c8204f79d4f3a83381b811b0adc"); // block 0
    uint256 fake_target_2 = uint256S("0xce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91"); // block 1
    uint256 fake_target_3 = uint256S("0x72e36f3fcdf98d3625dfe03f28a914c513b913231e479d53fc22e5e46cf5b585"); // block 17
    uint256 fake_target_4 = uint256S("0xf6f565e58812f89d8ea2aca296b9934ea82918f5bd443312af90b0be1465dbd1"); // block 510
    uint256 fake_replacement = uint256S("0x7777777777777777777777777777777777777777777777777777777777777777");
    if (block.GetHash() == fake_target_2 || block.GetHash() == fake_target_3 || block.GetHash() == fake_target_4) {
        // int nHeight = pindexPrev->nHeight + 1;
        printf("%s INPUT FAKE HEADER! %s\n", DateTimeStrFormat("%Y-%m-%d %H:%M:%S", GetTime()).c_str(), block.GetHash().ToString().c_str());
        if (fCheckPOW && !CheckProofOfWork(fake_replacement, block.nBits, consensusParams)) {
            return state.DoS(50, false, REJECT_INVALID, "high-hash (fake)", false, "proof of work failed (fake)");
        }
    } else if (fCheckPOW && !CheckProofOfWork(block.GetPoWHash_cached(), block.nBits, consensusParams)) {
        return state.DoS(50, false, REJECT_INVALID, "high-hash", false, "proof of work failed");
    }
```

decryp2kanon commented:
# TEST RESULT
- if you get some fake headers, daemon will bann the fake node. OK :heavy_check_mark: 
- if you recompile or delete `banlist.dat`, it syncs continuously well. OK :heavy_check_mark: 

printf
```
2020-04-28 20:21:27 CheckBlockH=c62f3395a8c53ebedb637bf80c83239e7bcd465133ef8c911a16c116ea9f3602
2020-04-28 20:21:27 AcceptBlock=c62f3395a8c53ebedb637bf80c83239e7bcd465133ef8c911a16c116ea9f3602
2020-04-28 20:21:27 INPUT FAKE HEADER! ce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91
2020-04-28 20:21:27 INPUT FAKE HEADER! 72e36f3fcdf98d3625dfe03f28a914c513b913231e479d53fc22e5e46cf5b585
2020-04-28 20:21:28 INPUT FAKE HEADER! ce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91
2020-04-28 20:21:28 INPUT FAKE HEADER! 72e36f3fcdf98d3625dfe03f28a914c513b913231e479d53fc22e5e46cf5b585
```

debug.log
```
2020-04-28 20:18:47 New outbound peer connected: version: 70015, blocks=4298257, peer=31
2020-04-28 20:21:26 Misbehaving: 216.155.157.233:34230 peer=1 (0 -> 50)
2020-04-28 20:21:26 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:26 Misbehaving: 216.155.157.233:34230 peer=1 (50 -> 100) BAN THRESHOLD EXCEEDED
2020-04-28 20:21:26 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:27 Misbehaving: 211.144.120.164:34230 peer=23 (0 -> 50)
2020-04-28 20:21:27 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:27 Misbehaving: 211.144.120.164:34230 peer=23 (50 -> 100) BAN THRESHOLD EXCEEDED
2020-04-28 20:21:27 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:27 Misbehaving: 45.77.192.211:34230 peer=0 (0 -> 50)
2020-04-28 20:21:27 ERROR: ProcessNewBlock: AcceptBlock FAILED (proof of work failed (fake))
2020-04-28 20:21:27 Misbehaving: 45.77.192.211:34230 peer=0 (50 -> 100) BAN THRESHOLD EXCEEDED
```

banned:
![image](https://user-images.githubusercontent.com/60179867/80534654-8bd3a280-89da-11ea-8202-6c19f4421f55.png)

decryp2kanon committed:
Revert "try: fake PoW test"

This reverts commit 360beb2bf071ca52633c3f44229957fba0ac61d9.

decryp2kanon commented:
problem when doing re-sync after banned
```
2020-04-28 21:15:57 UpdateTip: new best=32a228c2e43e5cd7470f6917eaf5970189587dfb78034a8dab16253801057988 height=280596 version=0x20000000 log2_work=40.269295 tx=322930 date='2019-09-09 07:06:20' progress=0.070321 cache=72.5MiB(527613txo)
2020-04-28 21:15:57 ERROR: ReadBlockFromDisk: Errors in block header at CBlockDiskPos(nFile=1, nPos=41745825)
2020-04-28 21:15:57 *** Failed to read block
```

```
2020-04-28 21:15:57 CheckBlockH=953ae5b3afc09c8f06effbf336cd4590c15d82c4318a42e2ab07902507f96f4e
2020-04-28 21:15:57 CheckBlockH=953ae5b3afc09c8f06effbf336cd4590c15d82c4318a42e2ab07902507f96f4e
2020-04-28 21:15:57 AcceptBlock=953ae5b3afc09c8f06effbf336cd4590c15d82c4318a42e2ab07902507f96f4e
2020-04-28 21:15:57 CheckBlockH=2608b1d17bef18ea96e88503687ee75d72c4267dcae0e6edd1fc3e1ca1b268ac
2020-04-28 21:15:57 CheckBlockH=2608b1d17bef18ea96e88503687ee75d72c4267dcae0e6edd1fc3e1ca1b268ac
2020-04-28 21:15:57 AcceptBlock=2608b1d17bef18ea96e88503687ee75d72c4267dcae0e6edd1fc3e1ca1b268ac
2020-04-28 21:15:57 CheckBlockH=2608b1d17bef18ea96e88503687ee75d72c4267dcae0e6edd1fc3e1ca1b268ac
2020-04-28 21:15:57 CheckBlockH=c0de58bc1a58aa72d54da04bf62c20722336927a6740dd5f3a7e6abddeba0f67
2020-04-28 21:15:57 CheckBlockH=fd2f0572c16266cc58f31042f2deab54cb90d7806e388f593d39d73c63f29013
2020-04-28 21:15:57 CheckBlockH=32a228c2e43e5cd7470f6917eaf5970189587dfb78034a8dab16253801057988
```

![image](https://user-images.githubusercontent.com/60179867/80538651-0ef7f700-89e1-11ea-977a-56751f8d5a3c.png)

- exit, and relaunch daemon fixes this.

OK

decryp2kanon committed:
try: fake PoW test, back to AcceptBlockHeader

decryp2kanon commented:
# ROLLBACK and TEST FAKE
98462b3

- different from this PR is, just print ERROR when first stage (downloading header)

![image](https://user-images.githubusercontent.com/60179867/80544397-f04b2d80-89eb-11ea-89f7-0903fdfb8ef4.png)

printf
```
2020-04-28 22:32:20 CheckBlockH=7d5eaec2dbb75f99feadfa524c78b7cabc1d8c8204f79d4f3a83381b811b0adc
2020-04-28 22:32:22 INPUT FAKE HEADER! ce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91
2020-04-28 22:32:26 INPUT FAKE HEADER! ce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91
```

debug.log
```
2020-04-28 22:32:20 16 addresses found from DNS seeds
2020-04-28 22:32:20 dnsseed thread exit
2020-04-28 22:32:21 New outbound peer connected: version: 70015, blocks=4299825, peer=0
2020-04-28 22:32:22 ERROR: AcceptBlockHeader: Consensus::CheckBlockHeader: ce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91, high-hash (fake), proof of work failed (fake) (code 16)
2020-04-28 22:32:22 Misbehaving: 51.89.173.178:34230 peer=0 (0 -> 50)
2020-04-28 22:32:22 ERROR: invalid header received
2020-04-28 22:32:23 New outbound peer connected: version: 70015, blocks=4299825, peer=1
2020-04-28 22:32:24 New outbound peer connected: version: 70015, blocks=4299825, peer=2
2020-04-28 22:32:26 ERROR: AcceptBlockHeader: Consensus::CheckBlockHeader: ce8a0df339f2edceb99c5325c95b2b0ae752e29de1193f6113549f0e1cae7c91, high-hash (fake), proof of work failed (fake) (code 16)
```

GOOD :heavy_check_mark:

decryp2kanon committed:
Revert "try: fake PoW test, back to AcceptBlockHeader"

This reverts commit 973270d90abe8fc6837055d84ac007b37911cd36.

cryptozeny commented:
check pow.cpp -debug=pow

decryp2kanon commented:
# offtopic
![image](https://user-images.githubusercontent.com/60179867/80552124-9dc83c00-8a00-11ea-9dcf-a70cd78f5c1f.png)

![image](https://user-images.githubusercontent.com/60179867/80552251-f7306b00-8a00-11ea-852c-21801204ab06.png)

in todo list, revert disabled many headers #78 almost no effect. (blue, debug40) 

# cons
- almost no effect
- increase traffic too much. 10x times bigger (4GB vs 50GB)

decryp2kanon commented:
# offtopic
no logging on pow.cpp and validation.cpp
- no effect
![image](https://user-images.githubusercontent.com/60179867/80606133-e3235280-8a6e-11ea-9d5e-bf89807af370.png)

decryp2kanon commented:
idea: test header, when testing blocks

decryp2kanon commented:
# reindex-chainstate 7:30
- only `CheckBlockH`
![image](https://user-images.githubusercontent.com/60179867/80658189-7173f480-8ac0-11ea-89c7-c67a06a430ca.png)

cryptozeny commented on src/validation.cpp:
```diff
@@ -3425,6 +3433,14 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
+    if (!CheckBlockHeader(block, state, chainparams.GetConsensus())) {
+        uint256 hash = block.GetHash();
+        return error("%s: Consensus::CheckBlockHeader: %s, %s", __func__, hash.ToString(), FormatStateMessage(state));
+    }
```
bool AcceptBlockHeader() = false;

decryp2kanon replied:
???

cryptozeny commented:
try make a better fake test

cryptozeny commented on src/validation.cpp:
```diff
@@ -3425,6 +3433,14 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
+    // came from AcceptBlockHeader (to here AcceptBlock)
+    if (!CheckBlockHeader(block, state, chainparams.GetConsensus())) {
+        uint256 hash = block.GetHash();
+        return error("%s: Consensus::CheckBlockHeader: %s, %s", __func__, hash.ToString(), FormatStateMessage(state));
```
checkblock not header

decryp2kanon replied:
its string.

decryp2kanon committed:
add: return false

decryp2kanon commented:
# offtopic

![image](https://user-images.githubusercontent.com/60179867/80777923-b3c13280-8ba1-11ea-8920-183a5a671a53.png)

```
2020-04-30 07:16:48 UpdateTip: new best=02e8ca1d5d149aaf51adb83c819ceae7503568d2e1b295d35a0787d32974c496 height=2210799 version=0x20000000 log2_work=42.51934 tx=2375296 date='2019-12-30 01:18:36' progress=0.514328 cache=716.1MiB(5313482txo)
2020-04-30 07:16:48 *** Disk space is low!
2020-04-30 07:36:49 socket sending timeout: 1201s
2020-04-30 07:36:49 socket sending timeout: 1201s
2020-04-30 07:36:49 socket sending timeout: 1201s
2020-04-30 07:36:49 socket sending timeout: 1201s
2020-04-30 07:36:49 socket sending timeout: 1201s
2020-04-30 07:36:49 socket sending timeout: 1201s
2020-04-30 07:36:49 socket sending timeout: 1201s
2020-05-01 02:48:04 GUI: QXcbWindow: Unhandled client message: "_COMPIZ_TOOLKIT_ACTION"
2020-05-01 02:48:18 ERROR: ProcessNewBlock: ActivateBestChain failed
2020-05-01 02:48:19 tor: Thread interrupt
2020-05-01 02:48:19 Shutdown: In progress...
2020-05-01 02:48:19 addcon thread exit
2020-05-01 02:48:19 torcontrol thread exit
2020-05-01 02:48:19 opencon thread exit
2020-05-01 02:48:19 net thread exit
2020-05-01 02:48:19 msghand thread exit
2020-05-01 02:48:19 scheduler thread interrupt
2020-05-01 02:48:19 Dumped mempool: 4e-06s to copy, 0.003703s to dump
2020-05-01 02:48:41 Shutdown: done
```

```
2020-04-30 07:16:48 CheckBlockH=4a3a208edd2c803d923c9cb47e9dcf625d5e8ee0d95fee20e6929ce11808b514
2020-04-30 07:16:48 AcceptBlock=4a3a208edd2c803d923c9cb47e9dcf625d5e8ee0d95fee20e6929ce11808b514
2020-04-30 07:16:48 CheckBlockH=4a3a208edd2c803d923c9cb47e9dcf625d5e8ee0d95fee20e6929ce11808b514
2020-04-30 07:16:48 CheckBlockH=210f49789bd5e49b3bd71c920e25d881e6738c75d5e9698b45995a2db93c9d72
2020-04-30 07:16:48 AcceptBlock=210f49789bd5e49b3bd71c920e25d881e6738c75d5e9698b45995a2db93c9d72
2020-04-30 07:16:48 CheckBlockH=210f49789bd5e49b3bd71c920e25d881e6738c75d5e9698b45995a2db93c9d72
2020-04-30 07:16:48 CheckBlockH=70b3e9d30718afccdb65ebb879907e0a8030f17e2d8d942b65744dd45c6c7989
2020-04-30 07:16:48 AcceptBlock=70b3e9d30718afccdb65ebb879907e0a8030f17e2d8d942b65744dd45c6c7989
2020-04-30 07:16:48 CheckBlockH=70b3e9d30718afccdb65ebb879907e0a8030f17e2d8d942b65744dd45c6c7989
2020-04-30 07:16:48 CheckBlockH=02e8ca1d5d149aaf51adb83c819ceae7503568d2e1b295d35a0787d32974c496
2020-04-30 07:16:48 AcceptBlock=02e8ca1d5d149aaf51adb83c819ceae7503568d2e1b295d35a0787d32974c496
2020-04-30 07:16:48 CheckBlockH=02e8ca1d5d149aaf51adb83c819ceae7503568d2e1b295d35a0787d32974c496
```

decryp2kanon committed:
tests: CheckProofOfWork mostly meaningless...

run tests
$ ./src/test/test_sugarchain test_bitcoin --log_level=test_suite --run_test=blockencodings_tests

decryp2kanon committed:
Revert "tests: CheckProofOfWork mostly meaningless..."

This reverts commit d8091a2a8c5952a87ef80dba4d175323bd243ad8.

decryp2kanon committed:
tests: BOOST_CHECK: CheckBlockHeader

decryp2kanon committed:
remove: printf

decryp2kanon committed:
copy: validation_block_tests.cpp
make -j4 && ./src/test/test_sugarchain test_bitcoin --log_level=test_suite --run_test=validation_fakeheader_tests

decryp2kanon committed:
fake header generater. test not yet
TODO: add BOOST_CHECK into loop

decryp2kanon committed:
Revert "tests: BOOST_CHECK: CheckBlockHeader"

This reverts commit 4ae25fd389a9bb8a40bcfc478789dde3225ef9fe.

decryp2kanon commented:
# fake header generator

```
Entering test suite "validation_fakeheader_tests"
Entering test case "processnewblockheaders_fake75"
2020-05-03 02:21:20 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 02:21:20 1/10 (9) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:21:21 2/10 (8) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:21:21 3/10 (7) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:21:21 4/10 (6) 8cab68e152d580b84d7b8ea0fce15f4aa98c83e64995729dd6dc8a4c93f653b8
2020-05-03 02:21:21 5/10 (5) 5bcd8043a8c200ed3d9f7c028c43ff45c105aac0deff2be272b59d7e425478c8
2020-05-03 02:21:21 6/10 (4) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:21:21 7/10 (3) fb268ace92940b9bde56b88f005dc836eb0ab18368a0bca46565ee972685788e
2020-05-03 02:21:21 8/10 (2) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:21:21 9/10 (1) 1b9ab8d89412a136460cdd46907e537d7e60ebb4e019c13d00ab593e656674df
2020-05-03 02:21:22 10/10 (0) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:21:22 PNBH=44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
Leaving test case "processnewblockheaders_fake75"; testing time: 1419986mks
Entering test case "processnewblockheaders_fake_zero"
2020-05-03 02:21:22 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 02:21:22 1/10 (9) d2f737b3f0a058eafc1a010613adce95e4e3b9b031493d177265ae60abc7b63f
2020-05-03 02:21:22 2/10 (8) 9c857da3e5ad6ef12be76343f875fa26b105767608d8b263394f537860d5298a
2020-05-03 02:21:22 3/10 (7) 831cc4bfe6fda1623e412db661f875fb7f3870d7214213e5993cb7febbd3c0e1
2020-05-03 02:21:22 4/10 (6) de4924d27d57249fc2efd1cc35e8f5a93b11edb49cf179b382ec338f54989698
2020-05-03 02:21:22 5/10 (5) 46e4de33585fdc2710e3c1602bc2bb4cde7a93d016f4f45152e9562f594742c9
2020-05-03 02:21:22 6/10 (4) 3fb78cd9203d4befe2d2653ad5bd39f1f2c4215fe590a949c0d836bd257b99d8
2020-05-03 02:21:22 7/10 (3) 326168bb77adf619eb8d075bde27411e10fdb11518f03b73ad9d83cc380b6830
2020-05-03 02:21:22 8/10 (2) 671bd162ff9c86a69a26fe070ef551ab660405fb8c7a4ef0732658a35c3e1ac3
2020-05-03 02:21:23 9/10 (1) 6783cfa641b8b503acc7fe7dba4587cb9ac896aeacf2906f3875fe5a79c4f22b
2020-05-03 02:21:23 10/10 (0) 792061829c34ef414320c55d1d959b5eeef7ba5e8563fd0196f57812e5960c94
2020-05-03 02:21:23 PNBH=d2f737b3f0a058eafc1a010613adce95e4e3b9b031493d177265ae60abc7b63f
2020-05-03 02:21:23 PNBH=9c857da3e5ad6ef12be76343f875fa26b105767608d8b263394f537860d5298a
2020-05-03 02:21:23 PNBH=831cc4bfe6fda1623e412db661f875fb7f3870d7214213e5993cb7febbd3c0e1
2020-05-03 02:21:23 PNBH=de4924d27d57249fc2efd1cc35e8f5a93b11edb49cf179b382ec338f54989698
2020-05-03 02:21:23 PNBH=46e4de33585fdc2710e3c1602bc2bb4cde7a93d016f4f45152e9562f594742c9
2020-05-03 02:21:23 PNBH=3fb78cd9203d4befe2d2653ad5bd39f1f2c4215fe590a949c0d836bd257b99d8
2020-05-03 02:21:23 PNBH=326168bb77adf619eb8d075bde27411e10fdb11518f03b73ad9d83cc380b6830
2020-05-03 02:21:23 PNBH=671bd162ff9c86a69a26fe070ef551ab660405fb8c7a4ef0732658a35c3e1ac3
2020-05-03 02:21:23 PNBH=6783cfa641b8b503acc7fe7dba4587cb9ac896aeacf2906f3875fe5a79c4f22b
2020-05-03 02:21:23 PNBH=792061829c34ef414320c55d1d959b5eeef7ba5e8563fd0196f57812e5960c94
Leaving test case "processnewblockheaders_fake_zero"; testing time: 1115621mks
Leaving test suite "validation_fakeheader_tests"
Leaving test suite "Sugarchain Test Suite"

*** No errors detected
```

NICE :+1:

decryp2kanon commented:
```
Entering test case "processnewblockheaders_fake_75"
2020-05-03 02:45:29 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 02:45:29 1/10 (9) 44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-03 02:45:30 2/10 (8) c2a78b86597c2740224a4c30ed8b2b2d147eeddd6c40ded511dd3fed320aa52d
2020-05-03 02:45:30 3/10 (7) 20d5a144b50dab915247f4fbcad1bf532e7563254222f89e2da208d44e2b006f
2020-05-03 02:45:30 4/10 (6) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:45:30 5/10 (5) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:45:30 6/10 (4) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:45:30 7/10 (3) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:45:30 8/10 (2) 05f1ee9a9b259b08ce78762fbd5660cc6315bd75f812b434ec96e65b2a59e9a9
2020-05-03 02:45:30 9/10 (1) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:45:30 10/10 (0) ba3e6220bb16faabc9c8109d5fa8351a70d53c0c13a64317838b2597b29216aa
2020-05-03 02:45:30 PNBH=44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f (9)
2020-05-03 02:45:30 PNBH=c2a78b86597c2740224a4c30ed8b2b2d147eeddd6c40ded511dd3fed320aa52d (8)
2020-05-03 02:45:30 PNBH=20d5a144b50dab915247f4fbcad1bf532e7563254222f89e2da208d44e2b006f (7)
2020-05-03 02:45:30 PNBH=756e1269433972bd00d5b404999fba9be08dedd2ea4500e992dddc75f12f6907 (failed)
```

decryp2kanon commented:
```
Entering test case "processnewblockheaders_fake_75"
2020-05-03 02:51:47 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 02:51:47 1/10 (9) 44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-03 02:51:47 2/10 (8) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:51:47 3/10 (7) b1ad2fdda85a7c911e2a96814c301119ad309a72ff8192fdfb2a7916c6489bd8
2020-05-03 02:51:47 4/10 (6) 66cf0c185c8119c4b6e1c03689d15c951dc0b113a85e33f8ddfc9973d35485fa
2020-05-03 02:51:47 5/10 (5) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:51:48 6/10 (4) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:51:48 7/10 (3) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:51:48 8/10 (2) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:51:48 9/10 (1) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:51:48 10/10 (0) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 02:51:48 PNBH=44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-03 02:51:48 PNBH=c2a78b86597c2740224a4c30ed8b2b2d147eeddd6c40ded511dd3fed320aa52d
2020-05-03 02:51:48 (false)PNBH=b1ad2fdda85a7c911e2a96814c301119ad309a72ff8192fdfb2a7916c6489bd8
Leaving test case "processnewblockheaders_fake_75"; testing time: 1344065mks
Entering test case "processnewblockheaders_fake_zero"
2020-05-03 02:51:48 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 02:51:48 1/10 (9) d2f737b3f0a058eafc1a010613adce95e4e3b9b031493d177265ae60abc7b63f
2020-05-03 02:51:48 2/10 (8) 9c857da3e5ad6ef12be76343f875fa26b105767608d8b263394f537860d5298a
2020-05-03 02:51:48 3/10 (7) 831cc4bfe6fda1623e412db661f875fb7f3870d7214213e5993cb7febbd3c0e1
2020-05-03 02:51:48 4/10 (6) de4924d27d57249fc2efd1cc35e8f5a93b11edb49cf179b382ec338f54989698
2020-05-03 02:51:48 5/10 (5) 46e4de33585fdc2710e3c1602bc2bb4cde7a93d016f4f45152e9562f594742c9
2020-05-03 02:51:49 6/10 (4) 3fb78cd9203d4befe2d2653ad5bd39f1f2c4215fe590a949c0d836bd257b99d8
2020-05-03 02:51:49 7/10 (3) 326168bb77adf619eb8d075bde27411e10fdb11518f03b73ad9d83cc380b6830
2020-05-03 02:51:49 8/10 (2) 671bd162ff9c86a69a26fe070ef551ab660405fb8c7a4ef0732658a35c3e1ac3
2020-05-03 02:51:49 9/10 (1) 6783cfa641b8b503acc7fe7dba4587cb9ac896aeacf2906f3875fe5a79c4f22b
2020-05-03 02:51:49 10/10 (0) 792061829c34ef414320c55d1d959b5eeef7ba5e8563fd0196f57812e5960c94
2020-05-03 02:51:49 PNBH=d2f737b3f0a058eafc1a010613adce95e4e3b9b031493d177265ae60abc7b63f
2020-05-03 02:51:49 PNBH=9c857da3e5ad6ef12be76343f875fa26b105767608d8b263394f537860d5298a
2020-05-03 02:51:49 PNBH=831cc4bfe6fda1623e412db661f875fb7f3870d7214213e5993cb7febbd3c0e1
2020-05-03 02:51:49 PNBH=de4924d27d57249fc2efd1cc35e8f5a93b11edb49cf179b382ec338f54989698
2020-05-03 02:51:49 PNBH=46e4de33585fdc2710e3c1602bc2bb4cde7a93d016f4f45152e9562f594742c9
2020-05-03 02:51:49 PNBH=3fb78cd9203d4befe2d2653ad5bd39f1f2c4215fe590a949c0d836bd257b99d8
2020-05-03 02:51:49 PNBH=326168bb77adf619eb8d075bde27411e10fdb11518f03b73ad9d83cc380b6830
2020-05-03 02:51:49 PNBH=671bd162ff9c86a69a26fe070ef551ab660405fb8c7a4ef0732658a35c3e1ac3
2020-05-03 02:51:49 PNBH=6783cfa641b8b503acc7fe7dba4587cb9ac896aeacf2906f3875fe5a79c4f22b
2020-05-03 02:51:49 PNBH=792061829c34ef414320c55d1d959b5eeef7ba5e8563fd0196f57812e5960c94
Leaving test case "processnewblockheaders_fake_zero"; testing time: 1144524mks
Leaving test suite "validation_fakeheader_tests"
Leaving test suite "Sugarchain Test Suite"

*** No errors detected
```
COOL :+1:

decryp2kanon committed:
tests: ProcessNewBlockHeaders: fake 50%

decryp2kanon commented:
```
Entering test case "processnewblockheaders_fake_50_newblock"
2020-05-03 05:21:02 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 05:21:03 1/10 (9) 44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-03 05:21:03 2/10 (8) c2a78b86597c2740224a4c30ed8b2b2d147eeddd6c40ded511dd3fed320aa52d
2020-05-03 05:21:03 3/10 (7) 20d5a144b50dab915247f4fbcad1bf532e7563254222f89e2da208d44e2b006f
2020-05-03 05:21:03 4/10 (6) 756e1269433972bd00d5b404999fba9be08dedd2ea4500e992dddc75f12f6907
2020-05-03 05:21:03 5/10 (5) f62dfeec9db5572db6fc62d86eb00d6576758da93e1335fd54e5a933cf981354
2020-05-03 05:21:03 6/10 (4) b9a02dd8d6d40cb3673972399d23f40a83c37c2ac6b099b1d7219a099960eefd
2020-05-03 05:21:03 7/10 (3) c1e8e2b5a95852daf8b8a6e27c065e92c44b46ff773e32c9a51ba771a6056444
2020-05-03 05:21:03 8/10 (2) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 05:21:03 9/10 (1) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 05:21:03 10/10 (0) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 05:21:03 PNB=44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-03 05:21:03 PNB=c2a78b86597c2740224a4c30ed8b2b2d147eeddd6c40ded511dd3fed320aa52d
2020-05-03 05:21:03 PNB=20d5a144b50dab915247f4fbcad1bf532e7563254222f89e2da208d44e2b006f
2020-05-03 05:21:03 PNB=756e1269433972bd00d5b404999fba9be08dedd2ea4500e992dddc75f12f6907
2020-05-03 05:21:03 PNB=f62dfeec9db5572db6fc62d86eb00d6576758da93e1335fd54e5a933cf981354
2020-05-03 05:21:03 PNB=b9a02dd8d6d40cb3673972399d23f40a83c37c2ac6b099b1d7219a099960eefd
2020-05-03 05:21:03 PNB=c1e8e2b5a95852daf8b8a6e27c065e92c44b46ff773e32c9a51ba771a6056444
2020-05-03 05:21:03 PNB=868a0769d18e705c3e72ea15bb03e2eb9cdd9bf463ae3f25ac0c21d965d10e11
2020-05-03 05:21:03 (false)PNB=1b9ab8d89412a136460cdd46907e537d7e60ebb4e019c13d00ab593e656674df
2020-05-03 05:21:03 (false)PNB=ba3e6220bb16faabc9c8109d5fa8351a70d53c0c13a64317838b2597b29216aa
Leaving test case "processnewblockheaders_fake_50_newblock"; testing time: 977115mks
Leaving test suite "validation_fakeheader_tests"
Leaving test suite "Sugarchain Test Suite"

*** No errors detected
```
OK

decryp2kanon committed:
tests: ProcessNewBlock

decryp2kanon commented:
```
Entering test suite "validation_fakeheader_tests"
Entering test case "processnewblockheaders_fake_50"
2020-05-03 06:26:11 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 06:26:11 1/10 (9) 44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-03 06:26:11 2/10 (8) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 06:26:12 3/10 (7) b1ad2fdda85a7c911e2a96814c301119ad309a72ff8192fdfb2a7916c6489bd8
2020-05-03 06:26:12 4/10 (6) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 06:26:12 5/10 (5) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 06:26:12 6/10 (4) 1a122d0afdd9c68265e9f9e9ef774a61799f110767a762fc153e0cb7febe5d74
2020-05-03 06:26:12 7/10 (3) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 06:26:12 8/10 (2) 05f1ee9a9b259b08ce78762fbd5660cc6315bd75f812b434ec96e65b2a59e9a9
2020-05-03 06:26:12 9/10 (1) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 06:26:12 10/10 (0) ba3e6220bb16faabc9c8109d5fa8351a70d53c0c13a64317838b2597b29216aa
2020-05-03 06:26:12 PNBH=44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-03 06:26:12 PNBH=c2a78b86597c2740224a4c30ed8b2b2d147eeddd6c40ded511dd3fed320aa52d
2020-05-03 06:26:12 (false)PNBH=b1ad2fdda85a7c911e2a96814c301119ad309a72ff8192fdfb2a7916c6489bd8
Leaving test case "processnewblockheaders_fake_50"; testing time: 1250963mks
Entering test case "processnewblockheaders_fake_zero"
2020-05-03 06:26:12 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 06:26:13 1/10 (9) d2f737b3f0a058eafc1a010613adce95e4e3b9b031493d177265ae60abc7b63f
2020-05-03 06:26:13 2/10 (8) 9c857da3e5ad6ef12be76343f875fa26b105767608d8b263394f537860d5298a
2020-05-03 06:26:13 3/10 (7) 831cc4bfe6fda1623e412db661f875fb7f3870d7214213e5993cb7febbd3c0e1
2020-05-03 06:26:13 4/10 (6) de4924d27d57249fc2efd1cc35e8f5a93b11edb49cf179b382ec338f54989698
2020-05-03 06:26:13 5/10 (5) 46e4de33585fdc2710e3c1602bc2bb4cde7a93d016f4f45152e9562f594742c9
2020-05-03 06:26:13 6/10 (4) 3fb78cd9203d4befe2d2653ad5bd39f1f2c4215fe590a949c0d836bd257b99d8
2020-05-03 06:26:13 7/10 (3) 326168bb77adf619eb8d075bde27411e10fdb11518f03b73ad9d83cc380b6830
2020-05-03 06:26:13 8/10 (2) 671bd162ff9c86a69a26fe070ef551ab660405fb8c7a4ef0732658a35c3e1ac3
2020-05-03 06:26:13 9/10 (1) 6783cfa641b8b503acc7fe7dba4587cb9ac896aeacf2906f3875fe5a79c4f22b
2020-05-03 06:26:14 10/10 (0) 792061829c34ef414320c55d1d959b5eeef7ba5e8563fd0196f57812e5960c94
2020-05-03 06:26:14 PNBH=d2f737b3f0a058eafc1a010613adce95e4e3b9b031493d177265ae60abc7b63f
2020-05-03 06:26:14 PNBH=9c857da3e5ad6ef12be76343f875fa26b105767608d8b263394f537860d5298a
2020-05-03 06:26:14 PNBH=831cc4bfe6fda1623e412db661f875fb7f3870d7214213e5993cb7febbd3c0e1
2020-05-03 06:26:14 PNBH=de4924d27d57249fc2efd1cc35e8f5a93b11edb49cf179b382ec338f54989698
2020-05-03 06:26:14 PNBH=46e4de33585fdc2710e3c1602bc2bb4cde7a93d016f4f45152e9562f594742c9
2020-05-03 06:26:14 PNBH=3fb78cd9203d4befe2d2653ad5bd39f1f2c4215fe590a949c0d836bd257b99d8
2020-05-03 06:26:14 PNBH=326168bb77adf619eb8d075bde27411e10fdb11518f03b73ad9d83cc380b6830
2020-05-03 06:26:14 PNBH=671bd162ff9c86a69a26fe070ef551ab660405fb8c7a4ef0732658a35c3e1ac3
2020-05-03 06:26:14 PNBH=6783cfa641b8b503acc7fe7dba4587cb9ac896aeacf2906f3875fe5a79c4f22b
2020-05-03 06:26:14 PNBH=792061829c34ef414320c55d1d959b5eeef7ba5e8563fd0196f57812e5960c94
Leaving test case "processnewblockheaders_fake_zero"; testing time: 1178676mks
Entering test case "processnewblockheaders_fake_50_newblock"
2020-05-03 06:26:14 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-03 06:26:14 1/10 (9) c3489f3cd013eab86f773845f21f2f57445e88075e04cafec2b9582cc5945560
2020-05-03 06:26:14 2/10 (8) f0b7f4a7dfe8be848542c55155aa83ccd4d082d8d770113df875d38a7c53f456
2020-05-03 06:26:14 3/10 (7) bbec496fbea1b610d67c48bbe633d7765e413b3e9bb2c061f85d8de55537e9da
2020-05-03 06:26:14 4/10 (6) 1dd502bb0e64e2732bb5da7b5cf699fab0ee167ca61b5c417868eeef827f0222
2020-05-03 06:26:14 5/10 (5) b3c1a57508f50cb5b51acf8be3fe1e9c46546f348bd76c42e4a0b5964563e195
2020-05-03 06:26:14 6/10 (4) deac6f8ae8b4fd329ed7bf70125a9ed51c948303ad47417d95a50e347d39cadb
2020-05-03 06:26:14 7/10 (3) 6422bcb3ae71504f50319deca6417e778d6fe6b199eb8a75cf58cbe152bb570f
2020-05-03 06:26:15 8/10 (2) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 06:26:15 9/10 (1) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 06:26:15 10/10 (0) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-03 06:26:15 PNB=c3489f3cd013eab86f773845f21f2f57445e88075e04cafec2b9582cc5945560
2020-05-03 06:26:15 PNB=f0b7f4a7dfe8be848542c55155aa83ccd4d082d8d770113df875d38a7c53f456
2020-05-03 06:26:15 PNB=bbec496fbea1b610d67c48bbe633d7765e413b3e9bb2c061f85d8de55537e9da
2020-05-03 06:26:15 PNB=1dd502bb0e64e2732bb5da7b5cf699fab0ee167ca61b5c417868eeef827f0222
2020-05-03 06:26:15 PNB=b3c1a57508f50cb5b51acf8be3fe1e9c46546f348bd76c42e4a0b5964563e195
2020-05-03 06:26:15 PNB=deac6f8ae8b4fd329ed7bf70125a9ed51c948303ad47417d95a50e347d39cadb
2020-05-03 06:26:15 PNB=6422bcb3ae71504f50319deca6417e778d6fe6b199eb8a75cf58cbe152bb570f
2020-05-03 06:26:15 PNB=cb65d5a10671ece13ed5fc6ae8f9757cc26f154b2b20130fc63b8efef86fd9a0
2020-05-03 06:26:15 (false)PNB=5208dfc1f4db584bb9bbda76f1da95b97cbc5778d713602e86d132a79acaa27f
2020-05-03 06:26:15 (false)PNB=f3bc8fef1c6dc85ef996f947d0c2fee26ceb73d0e5a03787f158876ec3227c53
Leaving test case "processnewblockheaders_fake_50_newblock"; testing time: 1136888mks
Leaving test suite "validation_fakeheader_tests"
Leaving test suite "Sugarchain Test Suite"

*** No errors detected
```
OK

cryptozeny commented on src/test/validation_fakeheader_tests.cpp:
```diff
@@ -0,0 +1,201 @@
+    pblock->vtx[0] = MakeTransactionRef(std::move(txCoinbase));
+
+    return pblock;
+    printf("%s pblock=%s\n", DateTimeStrFormat("%Y-%m-%d %H:%M:%S", GetTime()).c_str(), pblock->hashPrevBlock.ToString().c_str());
```
before return or remove

cryptozeny commented:
add !CheckProofOfWork

decryp2kanon committed:
comment

decryp2kanon committed:
remove: replace MakeBlock(root) with FinalizeBlock(Block(root))

decryp2kanon committed:
tests: pow

decryp2kanon committed:
fix: fake

decryp2kanon commented:
```
Running 4 test cases...
Entering test suite "Sugarchain Test Suite"
Entering test suite "validation_fakeheader_tests"
Entering test case "processnewblockheaders_fake_50"
2020-05-04 19:32:58 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-04 19:32:59 1/10 (9) 44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-04 19:32:59 2/10 (8) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:32:59 3/10 (7) b1ad2fdda85a7c911e2a96814c301119ad309a72ff8192fdfb2a7916c6489bd8
2020-05-04 19:32:59 4/10 (6) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:32:59 5/10 (5) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:32:59 6/10 (4) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:32:59 7/10 (3) fb268ace92940b9bde56b88f005dc836eb0ab18368a0bca46565ee972685788e
2020-05-04 19:33:00 8/10 (2) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:33:00 9/10 (1) 1b9ab8d89412a136460cdd46907e537d7e60ebb4e019c13d00ab593e656674df
2020-05-04 19:33:00 10/10 (0) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:33:00 PNBH=44d3fdc2f2f59f1f7e7431e9609fab7cd8d03fa27bfa4b0be1cdc96711353c5f
2020-05-04 19:33:00 (false)PNBH=e468b6fdda831111eae85087f4f4cd6503bacdd6bb1eb5ccc91b667e3ea61002
Leaving test case "processnewblockheaders_fake_50"; testing time: 1590692mks
Entering test case "processnewblockheaders_fake_zero"
2020-05-04 19:33:00 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-04 19:33:00 1/10 (9) d2f737b3f0a058eafc1a010613adce95e4e3b9b031493d177265ae60abc7b63f
2020-05-04 19:33:00 2/10 (8) 9c857da3e5ad6ef12be76343f875fa26b105767608d8b263394f537860d5298a
2020-05-04 19:33:00 3/10 (7) 831cc4bfe6fda1623e412db661f875fb7f3870d7214213e5993cb7febbd3c0e1
2020-05-04 19:33:00 4/10 (6) de4924d27d57249fc2efd1cc35e8f5a93b11edb49cf179b382ec338f54989698
2020-05-04 19:33:01 5/10 (5) 46e4de33585fdc2710e3c1602bc2bb4cde7a93d016f4f45152e9562f594742c9
2020-05-04 19:33:01 6/10 (4) 3fb78cd9203d4befe2d2653ad5bd39f1f2c4215fe590a949c0d836bd257b99d8
2020-05-04 19:33:01 7/10 (3) 326168bb77adf619eb8d075bde27411e10fdb11518f03b73ad9d83cc380b6830
2020-05-04 19:33:01 8/10 (2) 671bd162ff9c86a69a26fe070ef551ab660405fb8c7a4ef0732658a35c3e1ac3
2020-05-04 19:33:01 9/10 (1) 6783cfa641b8b503acc7fe7dba4587cb9ac896aeacf2906f3875fe5a79c4f22b
2020-05-04 19:33:01 10/10 (0) 792061829c34ef414320c55d1d959b5eeef7ba5e8563fd0196f57812e5960c94
2020-05-04 19:33:01 PNBH=d2f737b3f0a058eafc1a010613adce95e4e3b9b031493d177265ae60abc7b63f
2020-05-04 19:33:01 PNBH=9c857da3e5ad6ef12be76343f875fa26b105767608d8b263394f537860d5298a
2020-05-04 19:33:01 PNBH=831cc4bfe6fda1623e412db661f875fb7f3870d7214213e5993cb7febbd3c0e1
2020-05-04 19:33:01 PNBH=de4924d27d57249fc2efd1cc35e8f5a93b11edb49cf179b382ec338f54989698
2020-05-04 19:33:01 PNBH=46e4de33585fdc2710e3c1602bc2bb4cde7a93d016f4f45152e9562f594742c9
2020-05-04 19:33:01 PNBH=3fb78cd9203d4befe2d2653ad5bd39f1f2c4215fe590a949c0d836bd257b99d8
2020-05-04 19:33:01 PNBH=326168bb77adf619eb8d075bde27411e10fdb11518f03b73ad9d83cc380b6830
2020-05-04 19:33:01 PNBH=671bd162ff9c86a69a26fe070ef551ab660405fb8c7a4ef0732658a35c3e1ac3
2020-05-04 19:33:01 PNBH=6783cfa641b8b503acc7fe7dba4587cb9ac896aeacf2906f3875fe5a79c4f22b
2020-05-04 19:33:01 PNBH=792061829c34ef414320c55d1d959b5eeef7ba5e8563fd0196f57812e5960c94
Leaving test case "processnewblockheaders_fake_zero"; testing time: 1347372mks
Entering test case "processnewblockheaders_fake_50_newblock"
2020-05-04 19:33:01 0/10 (10) d567a9c891c7a47e6dd03f8006cb65b0d6406b5dc7b2c86d7a904815c394e1f1
2020-05-04 19:33:01 1/10 (9) c3489f3cd013eab86f773845f21f2f57445e88075e04cafec2b9582cc5945560
2020-05-04 19:33:01 2/10 (8) f0b7f4a7dfe8be848542c55155aa83ccd4d082d8d770113df875d38a7c53f456
2020-05-04 19:33:01 3/10 (7) bbec496fbea1b610d67c48bbe633d7765e413b3e9bb2c061f85d8de55537e9da
2020-05-04 19:33:01 4/10 (6) 1dd502bb0e64e2732bb5da7b5cf699fab0ee167ca61b5c417868eeef827f0222
2020-05-04 19:33:02 5/10 (5) b3c1a57508f50cb5b51acf8be3fe1e9c46546f348bd76c42e4a0b5964563e195
2020-05-04 19:33:02 6/10 (4) deac6f8ae8b4fd329ed7bf70125a9ed51c948303ad47417d95a50e347d39cadb
2020-05-04 19:33:02 7/10 (3) 6422bcb3ae71504f50319deca6417e778d6fe6b199eb8a75cf58cbe152bb570f
2020-05-04 19:33:02 8/10 (2) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:33:02 9/10 (1) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:33:02 10/10 (0) ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
2020-05-04 19:33:02 PNB=c3489f3cd013eab86f773845f21f2f57445e88075e04cafec2b9582cc5945560
2020-05-04 19:33:02 PNB=f0b7f4a7dfe8be848542c55155aa83ccd4d082d8d770113df875d38a7c53f456
2020-05-04 19:33:02 PNB=bbec496fbea1b610d67c48bbe633d7765e413b3e9bb2c061f85d8de55537e9da
2020-05-04 19:33:02 PNB=1dd502bb0e64e2732bb5da7b5cf699fab0ee167ca61b5c417868eeef827f0222
2020-05-04 19:33:02 PNB=b3c1a57508f50cb5b51acf8be3fe1e9c46546f348bd76c42e4a0b5964563e195
2020-05-04 19:33:02 PNB=deac6f8ae8b4fd329ed7bf70125a9ed51c948303ad47417d95a50e347d39cadb
2020-05-04 19:33:02 PNB=6422bcb3ae71504f50319deca6417e778d6fe6b199eb8a75cf58cbe152bb570f
2020-05-04 19:33:02 (false)PNB=3b6d3be8c4ef84f637241c263a2f7adde1d898e5d4631c1a52a0465203754f77
2020-05-04 19:33:02 (false)PNB=5208dfc1f4db584bb9bbda76f1da95b97cbc5778d713602e86d132a79acaa27f
2020-05-04 19:33:02 (false)PNB=f3bc8fef1c6dc85ef996f947d0c2fee26ceb73d0e5a03787f158876ec3227c53
Leaving test case "processnewblockheaders_fake_50_newblock"; testing time: 712042mks
Entering test case "processnewblockheaders_pow"
Leaving test case "processnewblockheaders_pow"; testing time: 17417mks
Leaving test suite "validation_fakeheader_tests"
Leaving test suite "Sugarchain Test Suite"

*** No errors detected
```
OK

decryp2kanon committed:
remove: debug comment out

decryp2kanon committed:
tests: copyright

decryp2kanon commented:
# DEBUG: accept two blocks

PNBH -> CheckBlockH -> PNB -> PNBH(again)
```
FIRST
2020-05-04 20:02:31 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40
2020-05-04 20:02:31 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40

2020-05-04 20:02:31 CheckBlockH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40
2020-05-04 20:02:31 CheckBlockH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40

2020-05-04 20:02:31 PNB=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40

2020-05-04 20:02:31 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40
2020-05-04 20:02:31 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40
2020-05-04 20:02:31 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40
2020-05-04 20:02:31 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40
2020-05-04 20:02:31 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40
2020-05-04 20:02:31 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40
2020-05-04 20:02:32 PNBH=e20f7000083a087ac96586799a79fb275175cf649152cde4ba324c6004f26c40

SECOND
2020-05-04 20:02:34 PNBH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2

2020-05-04 20:02:34 CheckBlockH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2
2020-05-04 20:02:34 CheckBlockH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2

2020-05-04 20:02:34 PNB=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2

2020-05-04 20:02:34 PNBH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2
2020-05-04 20:02:34 PNBH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2
2020-05-04 20:02:34 PNBH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2
2020-05-04 20:02:34 PNBH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2
2020-05-04 20:02:34 PNBH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2
2020-05-04 20:02:34 PNBH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2
2020-05-04 20:02:34 PNBH=509ae44602643e3bc0d61321fb4827cabb773ce5b3b51882de2647bc592cddd2
```

decryp2kanon commented:
check this process in moonlight ver.
```
2020-05-04 20:17:31 AcceptBlockH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNBH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 CheckBlockH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 CheckBlockH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNB=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNBH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNBH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNBH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNBH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNBH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNBH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
2020-05-04 20:17:31 PNBH=6f81e277996ca00db3449e18e9fd5c00e070663263228f29ae58287dca6a79d6
```

decryp2kanon committed:
debug

decryp2kanon commented:
moved to https://github.com/sugarchain-project/sugarchain/pull/122

closing
